pkg/cluster/fixetcd.go

```diff
@@ -0,0 +1,488 @@
+		if err != nil {
+			return nil
+		}
+
+		if b {
+			log.Infof("Deleting secret %s", s.Name)
+			err = secretcli.Delete(ctx, s.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func getPeerPods(pods []corev1.Pod, de *degradedEtcd, cluster string) (string, error) {
+	regNode, err := regexp.Compile(".master-[0-9]$")
+	if err != nil {
+		return "", err
+	}
+	regPod, err := regexp.Compile("etcd-" + cluster + "-[0-9A-Za-z]*-master-[0-9]$")
+	if err != nil {
+		return "", err
+	}
+
+	var peerPods string
+	for _, p := range pods {
+		if regNode.MatchString(p.Spec.NodeName) && regPod.MatchString(p.Name) && p.Name != de.Pod {
+			peerPods += p.Name + " "
+		}
+	}
+	return peerPods, nil
+}
+
+func fixPeers(ctx context.Context, log *logrus.Entry, de *degradedEtcd, pods *corev1.PodList, kubecli kubernetes.Interface, securitycli securityv1client.SecurityContextConstraintsInterface, cluster string) error {
+	peerPods, err := getPeerPods(pods.Items, de, cluster)
+	if err != nil {
+		return err
+	}
+
+	scc, _, _, sa, err := createPrivilegedServiceAccount(ctx, log, serviceAccountName, kubeServiceAccount, kubecli, securitycli)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Creating job %s", jobNameFixPeers)
+	b, err := kubecli.BatchV1().Jobs(nameSpaceEtcds).Create(ctx, &batchv1.Job{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      jobNameFixPeers,
+			Namespace: nameSpaceEtcds,
+			Labels:    map[string]string{"app": jobNameDataBackup},
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      jobNameFixPeers,
+					Namespace: nameSpaceEtcds,
+					Labels:    map[string]string{"app": jobNameDataBackup},
+				},
+				Spec: corev1.PodSpec{
+					RestartPolicy:      corev1.RestartPolicyOnFailure,
+					ServiceAccountName: serviceAccountName,
+					Containers: []corev1.Container{
+						{
+							Name:  "backup",
+							Image: image,
+							Command: []string{
+								"/bin/bash",
+								"-cx",
+								backupOrFixEtcd,
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Privileged: to.BoolPtr(true),
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "PEER_PODS",
+									Value: peerPods,
+								},
+								{
+									Name:  "DEGRADED_NODE",
+									Value: de.Node,
+								},
+								{
+									Name:  "FIX_PEERS",
+									Value: "true",
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "host",
+									MountPath: "/host",
+									ReadOnly:  false,
+								},
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: "host",
+							VolumeSource: corev1.VolumeSource{
+								HostPath: &corev1.HostPathVolumeSource{
+									Path: "/",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}, metav1.CreateOptions{
+		FieldManager: jobNameFixPeers,
+	})
+	if err != nil {
```

Comment:
If the function returns here because of an issue creating the job, the privileged service account will have been successfully created but won't be deleted. Should we be `defer` -ing the deletion of the service account and it's associated role and SCC right after creating them to keep this from happening?

Reply:
@kimorris27 That's a good point, thanks for calling that out. I'll test out adding defer to make sure the service account gets deleted in the event of an error.

Reply:
I've pushed a change that places all the other deletions in defer functions, testing locally it all works smoothly.